File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:

        if not root:
            return []
        if not root.left and not root.right:
            return [[root.val]]
        
        result, queue = list(), collections.deque([root])

        while queue:
            level, levelSize = list(), len(queue)
            
            for i in range(levelSize):

                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                level.append(node.val)
                
            result.append(level)
        
        return result
        

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        
        if(root.left == null && root.right == null) {return 1;}

        return this.dfs(root, root.val);
    }

    public int dfs(TreeNode node, int maxVal) {

        if(node == null) {return 0;}
        
        int count = (node.val >= maxVal) ? 1 : 0;
        maxVal = Integer.max(maxVal, node.val);
        int left = dfs(node.left, maxVal), right = dfs(node.right, maxVal);
        count += left;
        count += right;

        return count;
    }   
}
